[Reporting API] Test reports should go "Reporting-Endpoints" Endpoints

brentfulgham opened the pull request:
#### 8ae6f494b9976c5184a6e8058574f43d6ea77782
<pre>
[Reporting API] Test reports should go &quot;Reporting-Endpoints&quot; Endpoints
<a href="https://bugs.webkit.org/show_bug.cgi?id=244907">https://bugs.webkit.org/show_bug.cgi?id=244907</a>
&lt;rdar://problem/99666423&gt;

Reviewed by Chris Dumez.

Test reports should be sent to any endpoints provided in a &quot;Reporting-Endpoints&quot; header.
There also don&apos;t appear to be WPT that check this behavior. This patch implements
this part of the spec, and adds tests.

It also connects &apos;generateTestReport&apos; to WebKitLegacy so we can exercise the code in DumpRenderTree.

* LayoutTests/http/tests/security/contentSecurityPolicy: Rebasedlined tests.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub.html.sub.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub-expected: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub.html.sub.headers: Added.
* LayoutTests/platform/gtk/TestExpectations: Mark http-based WPT as expected to fail.
* LayoutTests/platform/mac-wk1/TestExpectations: Mark https-based WPT as expected to fail.
* Source/WebCore/Modules/reporting/Report.cpp:
(WebCore::Report::createReportFormDataForViolation):
* Source/WebCore/Modules/reporting/Report.h:
* Source/WebCore/Modules/reporting/ReportingScope.cpp:
(WebCore::ReportingScope::generateTestReport):
* Source/WebCore/Modules/reporting/TestReportBody.cpp:
(WebCore::TestReportBody::createReportFormDataForViolation const): Deleted.
* Source/WebCore/Modules/reporting/TestReportBody.h:
* Source/WebCore/loader/CrossOriginEmbedderPolicy.cpp:
(WebCore::sendCOEPInheritenceViolation):
(WebCore::sendCOEPCORPViolation):
* Source/WebCore/loader/CrossOriginOpenerPolicy.cpp:
(WebCore::sendViolationReportWhenNavigatingToCOOPResponse):
(WebCore::sendViolationReportWhenNavigatingAwayFromCOOPResponse):
* Source/WebCore/page/csp/CSPViolationReportBody.cpp:
(WebCore::CSPViolationReportBody::createReportFormDataForViolation const):
* Source/WebCore/page/csp/CSPViolationReportBody.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm
(-[WebFrame _generateTestReport:withGroup:]): Added.
* Source/WebKitLegacy/mac/WebView/WebFramePrivate.h
* Tools/DumpRenderTree/TestRunner.cpp
(generateTestReportCallback): Add support for &apos;generateTestReport&apos;.
(TestRunner::staticFunctions): Ditto.
* Tools/DumpRenderTree/TestRunner.h
* Tools/DumpRenderTree/mac/TestRunnerMac.mm
(TestRunner::generateTestReport): Added.
* Tools/DumpRenderTree/win/TestRunnerWin.cpp
(TestRunner::generateTestReport): Added stub.

Canonical link: <a href="https://commits.webkit.org/255845@main">https://commits.webkit.org/255845@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc0e2cc2dab65d25381cdb80d21eed51507c7db6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2968 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103417 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163736 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2984 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31218 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86107 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2125 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80209 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84051 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37613 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35465 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4038 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39343 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41276 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub.html:
```diff
@@ -0,0 +1,52 @@
+    }, "Generate Test Report");
+
+    promise_test(async t => {
+      await wait(3000);
```
Would be good if we could avoid such things.

brentfulgham replied:
Seems to work fine without the wait -- I'll remove them. We'll see what happens when I publish to WPT to see if other browsers fail.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub.html:
```diff
@@ -0,0 +1,52 @@
+    }, "Generate Test Report");
+
+    promise_test(async t => {
+      await wait(3000);
```
ditto.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/ReportingScope.cpp:
```diff
@@ -153,17 +153,30 @@ String ReportingScope::endpointURIForToken(const String& reportTo) const
     // https://w3c.github.io/reporting/#generate-test-report-command, step 7.1.10.
-    auto reportBody = TestReportBody::create(WTFMove(message));
-    notifyReportObservers(Report::create(reportBody->type(), WTFMove(reportURL), WTFMove(reportBody)));
+    if (auto* reportingClient = static_cast<ReportingClient*>(document)) {
```
It doesn't make sense that we cast document to its base class here. We should just null check document and use document instead of `reportingClient`. If the issue is that some Document overrides are private, then they should be made public. Why do virtual function calls when we already know the type?

brentfulgham replied:
I thought that was our typical pattern: Base class is public, children are private. But I have no personal objection to working as you propose here, so I'll make that change.

cdumez reviewed (approved):
LGTM with minor fixes.

Commit:
[Reporting API] Test reports should go "Reporting-Endpoints" Endpoints
https://bugs.webkit.org/show_bug.cgi?id=244907
<rdar://problem/99666423>

Reviewed by Chris Dumez.

Test reports should be sent to any endpoints provided in a "Reporting-Endpoints" header.
There also don't appear to be WPT that check this behavior. This patch implements
this part of the spec, and adds tests.

It also connects 'generateTestReport' to WebKitLegacy so we can exercise the code in DumpRenderTree.

* LayoutTests/http/tests/security/contentSecurityPolicy: Rebasedlined tests.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.https.sub.html.sub.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub-expected: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-honors-endpoint.http.sub.html.sub.headers: Added.
* LayoutTests/platform/gtk/TestExpectations: Mark http-based WPT as expected to fail.
* LayoutTests/platform/mac-wk1/TestExpectations: Mark https-based WPT as expected to fail.
* Source/WebCore/Modules/reporting/Report.cpp:
(WebCore::Report::createReportFormDataForViolation):
* Source/WebCore/Modules/reporting/Report.h:
* Source/WebCore/Modules/reporting/ReportingScope.cpp:
(WebCore::ReportingScope::generateTestReport):
* Source/WebCore/Modules/reporting/TestReportBody.cpp:
(WebCore::TestReportBody::createReportFormDataForViolation const): Deleted.
* Source/WebCore/Modules/reporting/TestReportBody.h:
* Source/WebCore/loader/CrossOriginEmbedderPolicy.cpp:
(WebCore::sendCOEPInheritenceViolation):
(WebCore::sendCOEPCORPViolation):
* Source/WebCore/loader/CrossOriginOpenerPolicy.cpp:
(WebCore::sendViolationReportWhenNavigatingToCOOPResponse):
(WebCore::sendViolationReportWhenNavigatingAwayFromCOOPResponse):
* Source/WebCore/page/csp/CSPViolationReportBody.cpp:
(WebCore::CSPViolationReportBody::createReportFormDataForViolation const):
* Source/WebCore/page/csp/CSPViolationReportBody.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm
(-[WebFrame _generateTestReport:withGroup:]): Added.
* Source/WebKitLegacy/mac/WebView/WebFramePrivate.h
* Tools/DumpRenderTree/TestRunner.cpp
(generateTestReportCallback): Add support for 'generateTestReport'.
(TestRunner::staticFunctions): Ditto.
* Tools/DumpRenderTree/TestRunner.h
* Tools/DumpRenderTree/mac/TestRunnerMac.mm
(TestRunner::generateTestReport): Added.
* Tools/DumpRenderTree/win/TestRunnerWin.cpp
(TestRunner::generateTestReport): Added stub.

Canonical link: https://commits.webkit.org/255845@main
